File: implementation/scipamato/core/sync/src/intTestDisabled/java/ch/difty/scipamato/core/sync/jobs/language/LanguageSyncConfigTest.java
package ch.difty.scipamato.core.sync.jobs.language;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.*;

import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Timestamp;

import org.jooq.DeleteConditionStep;
import org.jooq.TableField;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;
import org.springframework.batch.core.Job;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

import ch.difty.scipamato.core.db.tables.Language;
import ch.difty.scipamato.core.sync.jobs.SyncConfigTest;
import ch.difty.scipamato.publ.db.tables.records.LanguageRecord;

@SpringBootTest
class LanguageSyncConfigTest extends SyncConfigTest<LanguageRecord> {

    @Autowired
    private LanguageSyncConfig config;

    @Override
    protected Job getJob() {
        return config.syncLanguageJob();
    }

    @Override
    protected String selectSql() {
        return config.selectSql();
    }

    @Override
    protected TableField<LanguageRecord, Timestamp> lastSynchedField() {
        return config.lastSynchedField();
    }

    @Override
    protected DeleteConditionStep<LanguageRecord> getPseudoFkDcs() {
        return config.getPseudoFkDcs();
    }

    @Override
    protected String expectedJobName() {
        return "syncLanguageJob";
    }

    @Override
    protected String expectedSelectSql() {
        return "select \"public\".\"language\".\"code\", \"public\".\"language\".\"main_language\" from \"public\".\"language\"";
    }

    @Override
    protected TableField<LanguageRecord, Timestamp> expectedLastSyncField() {
        return ch.difty.scipamato.publ.db.tables.Language.LANGUAGE.LAST_SYNCHED;
    }

    @Test
    void makingEntity() throws SQLException {
        ResultSet rs = Mockito.mock(ResultSet.class);
        when(rs.getString(Language.LANGUAGE.CODE.getName())).thenReturn("de");
        when(rs.getBoolean(Language.LANGUAGE.MAIN_LANGUAGE.getName())).thenReturn(true);

        PublicLanguage pl = config.makeEntity(rs);

        assertThat(pl.getCode()).isEqualTo("de");
        assertThat(pl.getMainLanguage()).isTrue();
        assertThat(pl.getLastSynched()).isCloseTo("2016-12-09T06:02:13.000", 1000);

        verify(rs).getString(Language.LANGUAGE.CODE.getName());
        verify(rs).getBoolean(Language.LANGUAGE.MAIN_LANGUAGE.getName());

        verifyNoMoreInteractions(rs);
    }
}

File: core/core-sync/src/test/kotlin/ch/difty/scipamato/core/sync/jobs/language/LanguageSyncConfigTest.kt
package ch.difty.scipamato.core.sync.jobs.language

import ch.difty.scipamato.common.FrozenDateTimeService
import io.mockk.every
import io.mockk.mockk
import org.amshove.kluent.shouldBeEqualTo
import org.amshove.kluent.shouldBeInstanceOf
import org.amshove.kluent.shouldBeNull
import org.jooq.DSLContext
import org.jooq.SQLDialect
import org.jooq.impl.DSL
import org.jooq.tools.jdbc.MockConnection
import org.jooq.tools.jdbc.MockDataProvider
import org.junit.jupiter.api.Test
import org.springframework.batch.core.configuration.annotation.JobBuilderFactory
import org.springframework.batch.core.configuration.annotation.StepBuilderFactory
import java.sql.ResultSet
import java.sql.Timestamp
import javax.sql.DataSource

internal class LanguageSyncConfigTest {

    private val provider = mockk<MockDataProvider>()
    private val connection = MockConnection(provider)
    private val jooqCore: DSLContext = DSL.using(connection, SQLDialect.POSTGRES)
    private val jooqPublic: DSLContext = DSL.using(connection, SQLDialect.POSTGRES)
    private val coreDataSource = mockk<DataSource>()
    private val jobBuilderFactory = mockk<JobBuilderFactory>()
    private val stepBuilderFactory = mockk<StepBuilderFactory>()
    private val dateTimeService = FrozenDateTimeService()

    private val config = LanguageSyncConfig(
        jooqCore,
        jooqPublic,
        coreDataSource,
        jobBuilderFactory,
        stepBuilderFactory,
        dateTimeService
    )

    @Test
    fun jobName() {
        config.jobName shouldBeEqualTo "syncLanguageJob"
    }

    @Test
    fun publicWriter() {
        config.publicWriter() shouldBeInstanceOf LanguageItemWriter::class
    }

    @Test
    fun lastSynchedField() {
        config.lastSynchedField().toString() shouldBeEqualTo """"public"."language"."last_synched""""
    }

    @Test
    fun makingEntity() {
        val resultSet = mockk<ResultSet> {
            every { getString("code") } returns "mycode"
            every { getBoolean("main_language") } returns true
            every { getTimestamp("last_synched") } returns Timestamp.valueOf("2020-04-04 12:13:14")
            every { wasNull() } returns false
        }

        val language = config.makeEntity(resultSet)

        language.code shouldBeEqualTo "mycode"
        language.mainLanguage shouldBeEqualTo true
        language.lastSynched.toString() shouldBeEqualTo "2016-12-09 06:02:13.0"
    }

    @Test
    fun selectSql() {
        config.selectSql() shouldBeEqualTo
            """select "public"."language"."code", "public"."language"."main_language" from "public"."language""""
    }

    @Test
    fun pseudoFkDcs() {
        config.pseudoFkDcs.shouldBeNull()
    }
}
